internal/pkg/docker/orchestrator/orchestrator.go

```diff
@@ -185,6 +222,190 @@ func (a *runTaskAction) Do(o *Orchestrator) error {
+		if err != nil {
+			return fmt.Errorf("save iptables: %w", err)
+		}
+
+		err = o.docker.Exec(ctx, pauseContainer, io.Discard, "/bin/bash",
+			"-c", fmt.Sprintf(`echo %s %s >> /etc/hosts`, ip.String(), host.Name))
+		if err != nil {
+			return fmt.Errorf("update /etc/hosts: %w", err)
+		}
+
+		ip, err = increment(ip)
+		if err != nil {
+			return err
+		}
+
+		fmt.Printf("Created connection to %v:%v\n", host.Name, host.Port)
+	}
+
+	fmt.Printf("Finished setting up proxy connections\n\n")
+	return nil
+}
+
+// setupProxyConnection establishes an AWS SSM Port Forwarding connection to
+// host in pauseContainer via remoteContainer. Since this command is long-running,
+// errors that occur before discovering the epheremal port assigned to the
+// connection are returned. If an error occurs after the port is returned by
+// this function, it is reported as a runtime error to the main Orchestrator routine.
+func (o *Orchestrator) setupProxyConnection(ctx context.Context, pauseContainer, remoteContainer string, host Host) (string, error) {
+	errCh := make(chan error)
+	returned := make(chan struct{})
+	defer close(returned)
+
+	reportError := func(err error) {
+		select {
+		case errCh <- err:
+			// if setupProxyConnection() hasn't returned yet, report the error.
+			// this will never happen if setupProxyConnection() has already
+			// returned, as no goroutine will be able to read from errCh.
+		case <-returned:
+			// if setupProxyConnection() has already returned, we report
+			// this as a runtime error from the pause container
+			if o.curTaskID.Load() != orchestratorStoppedTaskID {
+				o.runErrs <- err
+			}
+		}
+	}
+
+	pr, pw := io.Pipe()
+	o.wg.Add(1)
+	go func() {
+		defer o.wg.Done()
+
+		err := o.docker.Exec(context.Background(), pauseContainer, pw, "aws", "ssm", "start-session",
+			"--target", remoteContainer,
+			"--document-name", "AWS-StartPortForwardingSessionToRemoteHost",
+			"--parameters", fmt.Sprintf(`{"host":["%s"],"portNumber":["%s"]}`, host.Name, host.Port))
+		if err != nil {
+			reportError(fmt.Errorf("proxy to %v:%v: %w", host.Name, host.Port, err))
+		}
+	}()
+
+	scanner := bufio.NewScanner(pr)
+	lines := make(chan string)
+
+	// read the output of exec, so the buffer doesn't
+	// fill up and block the ssm session process. until setupProxyConnection()
+	// returns, we'll send the output to lines.
+	o.wg.Add(1)
+	go func() {
+		defer o.wg.Done()
+		for scanner.Scan() {
+			line := scanner.Text()
+			select {
+			case lines <- line:
+			default:
+			}
+		}
+		if err := scanner.Err(); err != nil {
+			reportError(fmt.Errorf("process output for proxy to %v:%v: %w", host.Name, host.Port, err))
+		}
+	}()
+
+	var port string
+	for port == "" {
+		select {
+		case <-ctx.Done():
+			return "", ctx.Err()
+		case err := <-errCh:
+			return "", err
+		case line := <-lines:
+			// the line we want looks like:
+			// Port 61972 opened for sessionId mySessionId
```

Comment:
Can we use [describe-sessions](https://docs.aws.amazon.com/cli/latest/reference/ssm/describe-sessions.html) to check for the status instead of relying on output parsing 💭

Reply:
i don't think `describe-sessions` includes the local port it bound to 🤔 we _could_ explicitly pass the port we want it to bind to for every host, and then we wouldn't have to parse any output. the downside being, if we pick a port the customer's code wants to use, then we won't have a workaround for them.

Reply:
after some offline discussion, we decided to remove the output parsing and explicitly pass a port number to bind to using the `localPortNumber` parameter. we'll start at port 50000 and increment from there. we may have to introduce a flag in the future to all a customer to skip ports that they want their application to bind to, but we'll start without that flag.